### PR TITLE
perf(lsp): release unused documents

### PR DESCRIPTION
**Pull request by nayeemrmn**
The TS language service will call `documentRegistry.releaseDocumentWithKey()` for sources that aren't found in the graph. We couldn't benefit from that until #23259 because we used to pass _all_ known documents as roots.

Also, don't use `sourceFileCache` for LSP documents. There is a different `ts.SourceFile` API which is used by the document registry. Trying to reuse `host.getSourceFile()`/`sourceFileCache` was causing sources to be parsed twice I think. Use a separate `sourceTextCache`.

Like the VSCode TS server does, use a ref counter to count acquisitions of source files and only release when it decrements to 0. Currently this can only move between 1 and 0. But later when we have a `ts.LanguageService` object per-deno.json-scope, it will count how many scopes are currently interested in each source file.

Replace `TsServer::restart()` with a more targeted `TsServer::cleanup_semantic_cache()`. We don't want to lose these ref counts in the middle of the program.

### CODE REVIEW

File: cli/tsc/99_main_compiler.js [hunk 1]
```diff
@@ -1152,12 +1178,6 @@ delete Object.prototype.__proto__;
     debug("serverInit()");
   }
 
-  function serverRestart() {
-    languageService = ts.createLanguageService(host, documentRegistry);
-    isNodeSourceFileCache.clear();
```
Comment (nathanwhit) [inline cli/tsc/99_main_compiler.js:1157]: Does this break anything?

I'm not sure if we cleared this cache for a specific reason or if it's just "be safe and clear the cache" legacy code.

Comment (bartlomieju) [reply cli/tsc/99_main_compiler.js:1157]: Isn't this one used by `Deno: Restart Language Server` task in `vscode_deno`?

Comment (nayeemrmn) [reply cli/tsc/99_main_compiler.js:1157]: It shouldn't in practice, this checks if a path contains a `/node_modules/` component or something similar. It shouldn't change values as a result of caching or changing watched files.

Actually to be safe I'll delete it on config changes. Maybe it's affected for non-byonm npm resolvers.

Comment (nayeemrmn) [reply cli/tsc/99_main_compiler.js:1157]: `Deno: Restart Language Server` restarts this whole process.

**Review by nathanwhit — APPROVED** [review]
LGTM

**Pull request merged**